src/tool/EventTool.cpp

```diff
@@ -187,26 +196,76 @@ void EventTool::copyAction()
                 OffEvent* offEv = dynamic_cast<OffEvent*>(onEv->offEvent()->copy());
                 if (offEv) {
                     offEv->setOnEvent(onEv);
-                    copiedEvents->append(offEv);
+                    //copiedEvents->append(offEv);
+                    offEv->setTrack(copyTrack, false);
+                    offEv->setChannel(0, false);
+                    offEv->setFile(&copyFile);
+                    copyFile.channelEvents(0)->insert(offEv->midiTime(), offEv);
                 }
             }
         }
-        _mainWindow->copiedEventsChanged();
+        
+        QByteArray copyFileBytes = copyFile.toByteArray();
+        int MAX_COPY_SIZE = 1073741824; //1GB
+        if (copyFileBytes.size() > MAX_COPY_SIZE){
+            QMessageBox msgBox;
+            msgBox.setText("Copied data too large...");
+            msgBox.exec();
+            return;
+        }
+
+        if (!sharedMemory.isAttached()){
+            if (!sharedMemory.create(MAX_COPY_SIZE)){
+                QMessageBox msgBox;
+                msgBox.setText("Failed to copy items...");
+                msgBox.exec();
+                return;
+            }
+        }
+
+        sharedMemory.lock();
+        char* to = (char*) sharedMemory.data();
+        char* from = copyFileBytes.data();
+        memcpy(to, from, qMin(sharedMemory.size(), copyFileBytes.size()));
+        sharedMemory.unlock();
+
+        //_mainWindow->copiedEventsChanged();
     }
 }
 
 void EventTool::pasteAction()
 {
+    if (!sharedMemory.isAttached() && !sharedMemory.attach()){
+        return; //No data copied
+    }
+    sharedMemory.lock();
+    QByteArray pasted_file_raw = QByteArray();
+    pasted_file_raw.setRawData((char*)sharedMemory.constData(), sharedMemory.size());
+    
+    if(pasted_file_raw.size() == 0) {
+        sharedMemory.unlock();
+        return;
+    };
+    bool ok = true;
+    MidiFile pasted_file(pasted_file_raw, &ok, nullptr);
+
+    sharedMemory.unlock();
 
-    if (copiedEvents->size() == 0) {
+    if(!ok){
```

Comment:
Fine... But how it can be false? Okay...

Reply:
Probably should never be false, but since the MidiFile constructor already has a check-mechanism, why not use it